classes/script_metadata.php

```diff
@@ -209,18 +209,38 @@ public static function get_redactable_param_names(): array {
      *
      * @return string the request path for this profile.
      */
-    public static function get_request(): string {
-        global $SCRIPT, $ME, $CFG;
+    public static function get_normalised_relative_script_path(): string {
+        global $SCRIPT;
 
-        if (!isset($ME)) {
-            // If set, it will trim off the leading '/' to normalise web & cli requests.
-            $request = isset($SCRIPT) ? ltrim($SCRIPT, '/') : self::REQUEST_UNKNOWN;
-            return $request;
+        // This is from the previous iteration of the function.
+        if (!isset($SCRIPT)) {
```

Comment:
cam you please cover this with unit tests, the removal of the logic around $ME seems a bit risky to me, it might be fine but having the tests proves it and I think there is some edge cases here which we don't want to regress

This would be a good candidate for a dataProvider, see the existing tests for an example

Reply:
Definitely needs unit tests but how would they realistically cover the interaction between $SCRIPT and $ME?

Reply:
Currently get_normalised_relative_script_path is not a pure function as it has an input the value inside $SCRIPT, and it did previously also use $ME. I think that there is some weird edge cases where the value of $ME should be leveraged and this should have been covered by unit tests before so we should fix that and make sure they are covered. Off the top of my head one edge case to test for is when you go to a page which doesn't exist, like /does-not-exist then nginx served this as an error page which is /error/index.php and I think it needs both globals to correctly handle that.

As for the unit tests, there is two approaches and both would have the value of $SCRIPT as the first input, the value of $ME as the second, and the third would be the expected outcome
1) in the unit test you simply set $SCRIPT and $ME to what the data provider give it and then run the function. You should probably stash the original values and reset them so other later unit tests don't get weird side effects.
2) we remove the global dependency on $SCRIPT and $ME and make these explicit function parameters eg $request = script_metadata::get_normalised_relative_script_path($SCRIPT, $ME); - this one is the more proper approach but you'll find lots of examples of the first approach in various core unit tests

Reply:
It is now a pure function with some added heuristics for better normalisation.